tests/e2e/contracts.go

```diff
@@ -89,9 +94,159 @@ func ContractsTest(sc *RuntimeScenario, log *logging.Logger, conn *grpc.ClientCo
 		return fmt.Errorf("failed to decode contract result: %w", err)
 	}
 
-	if result["hello"]["greeting"] != fmt.Sprintf("hello e2e test (%d)", initialCounter) {
+	if result["hello"]["greeting"] != fmt.Sprintf("hello e2e test (%d)", counter) {
+		return fmt.Errorf("unexpected result from contract: %+v", result)
+	}
+	// Calling say_hello bumps the counter.
+	counter++
+
+	// Upload OAS20 contract code.
+	tb = ct.Upload(contracts.ABIOasisV1, contracts.Policy{Everyone: &struct{}{}}, oas20ContractCode).
+		SetFeeGas(1_000_000).
+		AppendAuthSignature(signer.Public(), nonce+3)
+	_ = tb.AppendSign(ctx, signer)
+	var uploadOas20 contracts.UploadResult
+	if err := tb.SubmitTx(ctx, &uploadOas20); err != nil {
+		return fmt.Errorf("failed to upload contract: %w", err)
+	}
+
+	// Instantiate OAS20 contract.
+	tb = ct.Instantiate(
+		uploadOas20.ID,
+		contracts.Policy{Everyone: &struct{}{}},
+		// This needs to conform to the contract API.
+		map[string]interface{}{
+			"instantiate": &oas20.Instantiate{
+				Name:     "OAS20 Test token",
+				Symbol:   "OAS20TEST",
+				Decimals: 2,
+				InitialBalances: []oas20.InitialBalance{
+					{
+						Address: types.NewAddress(signer.Public()),
+						Amount:  *quantity.NewFromUint64(10_000),
+					},
+				},
+			},
+		},
+		[]types.BaseUnits{},
+	).
+		SetFeeGas(1_000_000).
+		AppendAuthSignature(signer.Public(), nonce+4)
+	_ = tb.AppendSign(ctx, signer)
+	var instanceOas20 contracts.InstantiateResult
+	if err := tb.SubmitTx(ctx, &instanceOas20); err != nil {
+		return fmt.Errorf("failed to instantiate OAS20 contract: %w", err)
+	}
+
+	// Tansfer some tokens.
+	tb = ct.Call(
+		instanceOas20.ID,
+		&oas20.Request{
+			Transfer: &oas20.Transfer{
+				To:     testing.Charlie.Address,
+				Amount: *quantity.NewFromUint64(10),
+			},
+		},
+		[]types.BaseUnits{},
+	).
+		SetFeeGas(1_000_000).
+		AppendAuthSignature(signer.Public(), nonce+5)
+	_ = tb.AppendSign(ctx, signer)
+	if err := tb.SubmitTx(ctx, &rawResult); err != nil {
+		return fmt.Errorf("failed to call OAS20 transfer: %w", err)
+	}
+
+	// TODO: watch emitted event.
```

Comment:
I guess it would be great to have some more high-level wrapper for processing events (as currently you just get the raw tags) in the Go client-sdk and then specifically for contract events. Probably would need to be something generic with some per-module helpers for those specific events and then something for contracts on top of that.